### PR TITLE
issue 777416: Add SO_VMA_USER_DATA option for setsockopt()

### PR DESCRIPTION
**Pull request by igor-ivanov**
New VMA specific option allows to provide user defined data
in vma_completion_t as result of vma_poll().
By default socket descriptor is returned.
SO_VMA_USER_DATA option associated with SOL_SOCKET level.

Signed-off-by: Igor Ivanov <igor.ivanov.va@gmail.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/168/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-redmine-issue/51/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/168/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-redmine-issue/52/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/169/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/169/ for details (Mellanox internal link).

**Comment by rosenbaumalex** [comment]
+1

File: src/vma/sock/sockinfo_tcp.cpp [hunk 1]
```diff
@@ -3293,6 +3300,13 @@ int sockinfo_tcp::getsockopt_offload(int __level, int __optname, void *__optval,
 		return ret;
 	}
 
+	ret = sockinfo::getsockopt(__level, __optname, __optval, __optlen);
+	if (0 == ret) {
+		return 0;
+	}
+
+	ret = 0;
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_tcp.cpp]: ret was initialized to -1 
```code
int ret = -1
```
but here you override it with 0

```code
ret = 0;
```

File: src/vma/sock/sockinfo_udp.cpp [hunk 2]
```diff
@@ -1079,6 +1088,13 @@ int sockinfo_udp::getsockopt(int __level, int __optname, void *__optval, socklen
 	if (unlikely(m_b_closed) || unlikely(g_b_exit))
 		return ret;
 
+	ret = sockinfo::getsockopt(__level, __optname, __optval, __optlen);
+	if (0 == ret) {
+		return 0;
+	}
+
+	ret = 0;
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_udp.cpp]: ret was initialized as follows:
```code
int ret = orig_os_api.getsockopt(m_fd, __level, __optname, __optval, __optlen);
```
but here you override it with 
```code
ret = 0;
```

File: src/vma/sock/sockinfo_tcp.cpp (hunk 1 above)
Comment (igor-ivanov) [inline src/vma/sock/sockinfo_tcp.cpp]: @OphirMunk setsockopt looks ok but getsockopt requires changes.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-redmine-issue/53/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/170/ for details (Mellanox internal link).

**Comment by OphirMunk** [comment]
+1

**Comment by OphirMunk** [comment]
:+1:

**Pull request merged**